### PR TITLE
Define latencies below 150 ms as near real-time latencies.

### PR DESCRIPTION
**Pull request by maxsharabayko**
I propose defining latencies less than 150 ms as near real-time latencies.

These latencies can be used by musicians to play together in sync. And I guess the human ear or eye would mostly not notice any delay below 50 ms.
Might be worth doing a bit of research around the topic to refine the ranges though.
For instance, in this paper (https://ieeexplore.ieee.org/document/5540888) they conclude:
> ... the visual response latency ranges from 74ms to 106ms...

In another source (https://www.pubnub.com/blog/how-fast-is-realtime-human-perception-and-technology/) the following is defined:
> 300ms < game is unplayable
    150ms  < game play degraded 
    100ms < player performance affected
    50ms   > target performance
    13ms    > lower detectable limit

< 200 ms - comfortable latencies for conversation.

**_Note._** Migrated from https://github.com/fiestajetsam/draft-gruessing-moq-requirements/pull/64

_(Sorry for proposing this close to the last call)._

### CODE REVIEW

File: draft-ietf-mops-streaming-opcons.md [hunk 1]
```diff
@@ -553,10 +553,20 @@ Streaming media latency refers to the "glass-to-glass" time duration, which is t
...
-- low-latency live     (less than 10 seconds)
-- non-low-latency live (10 seconds to a few minutes)
-- on-demand            (hours or more)
+- near realtime latency (less than 150 milliseconds)
+- ultra low-latency     (less than 1 second)
+- low-latency live      (less than 10 seconds)
+- non-low-latency live  (10 seconds to a few minutes)
+- on-demand             (hours or more)
+
+## Near Realtime Latency
```
Comment (GrumpyOldTroll) [inline draft-ietf-mops-streaming-opcons.md:562]: My first instinct for the explanation section would be aiming for some text that hits the following points:

 - explaining that this latency category is driven by interactivity requirements
 - that a variety of interactive applications exist (and are envisioned) and they have a variety of different latency targets
 - that the latency requirements are driven by a broader system that includes the application's interactivity feedback mechanism and sensitivity to timing of the response
 - a full review of the use cases and techniques for satisfying them is out of scope for this doc, even though these applications sometimes do incorporate elements of streaming media with hard latency requirements.

(As long as we say something along these lines, I'm fine with also listing a few known example applications, especially if they have references for further reading.)

It also could be worth mentioning that some of the points in the next category (ultra low latency) are applicable here as well, or moving some of the text to this section, if we add it.

**Review by GrumpyOldTroll — CHANGES_REQUESTED** [review]
Thanks for the PR, @maxsharabayko .  I believe we have not yet entered IETF last call, so I don't think it's too late for comments like this, and proposing text is much appreciated!

My overall response as an individual:

I think listing out the use cases for this category of latency is probably out of scope for this doc.

However, declaring that this category of latency is a relevant boundary that matters for some applications, and particularly the claim that "< 200ms is comfortable for conversation" is perhaps a good addition... Do you have a reference for that number?  If that can be supported, maybe the boundary should be at 200ms for the higher end of "near realtime"?

**Comment by maxsharabayko** [comment]
> However, declaring that this category of latency is a relevant boundary that matters for some applications, and particularly the claim that "< 200ms is comfortable for conversation" is perhaps a good addition... Do you have a reference for that number? If that can be supported, maybe the boundary should be at 200ms for the higher end of "near realtime"?

I lost the original reference. Will try to dig it out.
Meanwhile here are some other interesting studies on the topic.

---

> At 100 ms of latency, the users start talking on top of each other, and at 300 ms, the conversation becomes impossible to follow.

Source: [Impact of Packet Loss, Jitter, and Latency on VoIP](https://netbeez.net/blog/impact-of-packet-loss-jitter-and-latency-on-voip/)

---

In telesurgery:

> In both exercises, latencies ≤300 ms were judged to be safe by all and 400-500 ms were accepted by 66-75 % of subjects. Less than 20 % of subjects accepted delays ≥800 ms.
(...) Latencies ≤200 ms are ideal for telesurgery; 300 ms is also suitable; 400-500 ms may be acceptable but are already tiring; and 600-700 ms are difficult to deal with and only acceptable for low risk and simple procedures. Surgery is quite difficult at 800-1,000 ms, telementoring would be a better choice in this case.

Source: [Determination of the latency effects on surgical performance and the acceptable latency levels in telesurgery using the dV-Trainer(®) simulator](https://pubmed.ncbi.nlm.nih.gov/24671353/)

---

No certain numbers are presented, but a good study about the subjective perception of latencies in conversations.

> (...) But one thing that impels speakers is that overlong pauses between turns seem to have semiotic significance, especially when the first turn requires a response, as in the case of questions, offers, requests, and so on. For example, when you invite someone to a party, a silence of, say, one second, might be indicative that the response might not be what you hoped for. Or consider how the caller C in this telephone call interprets a 1.86 secs pause as a negative answer to his own question which he himself voices (Levinson, 1995):
- C: So um (0.2) I was wondering would you be in your office (0.62) on Monday (0.42) by any chance?
- (1.86)
- C: Probably not

Source: [Conversational expectations get revised as response latencies unfold](https://pure.mpg.de/rest/items/item_3032071_5/component/file_3249633/content)

File: draft-ietf-mops-streaming-opcons.md [hunk 2]
```diff
@@ -553,10 +553,20 @@ Streaming media latency refers to the "glass-to-glass" time duration, which is t
...
+
+Near realtime delivery of media is defined here as having a glass-to-glass (or ear-to-ear) delay target under 150 milliseconds.
+The human ear or eye would mostly not notice or would be on the edge of noticing such latencies.
+These latencies are targeted in various telepresence use cases, such as remote control of drones and autonomous vehicles (e.g. teledriving), remote surgery. For musicians to play together in sync the ideal delay would be below 50 ms.
```
[the remark this answers is not in the document]
Comment (maxsharabayko) [reply draft-ietf-mops-streaming-opcons.md:566]: I assume there should anyway be some orchestration for this kind of scenario. Like you say, even for an orchestra to play in sync while being in the same room is tricky because of the speed of sound.
There should be a common synchronization source, musicians follow it to play along. And then maybe some post-processing to sync all musicians according to unequal delays.
The exact number is subjective. But I guess the truth is somewhere around 50 ms. As always, the lower latency, the better :)

File: draft-ietf-mops-streaming-opcons.md [hunk 3]
```diff
@@ -553,10 +553,20 @@ Streaming media latency refers to the "glass-to-glass" time duration, which is t
 
 Streaming media can be usefully categorized according to the application's latency requirements into a few rough categories:
 
-- ultra low-latency    (less than 1 second)
-- low-latency live     (less than 10 seconds)
-- non-low-latency live (10 seconds to a few minutes)
-- on-demand            (hours or more)
+- near realtime latency (less than 150 milliseconds)
```
[the remark this answers is not in the document]
Comment (maxsharabayko) [reply draft-ietf-mops-streaming-opcons.md:556]: Totally agree. I would borrow some parts of your suggestion. 😃

**Comment by maxsharabayko** [comment]
@GrumpyOldTroll Thank you for the review. I wonder what would be the general take on this proposal?
I mean, we can refine it a lot and make a candy out of it. It makes sense if there is a consensus on including the section in the doc.
Or should it happen the other way round where candy is prepared first, and then it is being sold to the working group? 😄

**Comment by SpencerDawkins** [comment]
> I guess I'd like to hear from @acbegen and @SpencerDawkins first, they might have a different take from me.

You rang? 

I have two theories:

- I was one of the people who was asking recently about what seemed to be "generous" latency budgets, but was satisfied with the PR that stopped talking about videoconferencing, because that's technically not exactly "streaming video". Are we thinking streaming operators are going to be consulting this document for guidance about videoconferencing? 
- If we're going forward for this discussion, we probably need to do that on the mailing list rather than in Github, but that's a better question for @ldaigle and @squarooticus than for me to answer. 

And don't get me wrong - I think the IETF does owe itself a draft that has a broader and better description of latency (even application-level latency) than we've got in this draft, but the question is whether we need to hold that document until we agree on a description that would also be useful for "Media over QUIC", among other things.

**Comment by GrumpyOldTroll** [comment]
Maybe what I meant to suggest was something more like a sub-secton under ultra-low latency in 4.1 to say "these things are out of scope" harder and more concretely than we currently say it.  For example, something like this:

----

OLD ([last sentence in 4.1](https://htmlpreview.github.io/?https://github.com/ietf-wg-mops/draft-ietf-mops-streaming-opcons/blob/gh-pages/draft-ietf-mops-streaming-opcons.html#section-4.1-7)):
Applications with further-specialized latency requirements are out of scope for this document.

NEW:
### Near-Realtime Latency

Some internet applications that incorporate media streaming have specific interactivity or control-feedback requirements that drive much lower glass to glass media latency targets than one second.  These include videoconferencing or voice calls, remote video game play, remote control of hardware platforms like drones, vehicles, or surgical robots, and many other envisioned or deployed interactive applications. [refs if we have them]

Applications with latency targets in these regimes are out of scope for this document.

----

I'm not sure it's critical, but it might help to acknowledge more explicitly that lower latency regimes do exist where we declare them out of scope, and maybe to do so in a way that'll appear in the table of contents, so that the people looking for this will more quickly figure out they should be looking somewhere else.

It also seems potentially helpful to give the lower latency domain a name.  I'm slightly disturbed at the way moq is tempted to use these categories even though many of them want to be operating in a domain that's an order of magnitude or 2 finer-grained than "ultra low", and having a name for this category seems a possibly helpful way to avoid confusion with "ultra low" as used in things like cmaf, the way we mean to use it here.

**Comment by maxsharabayko** [comment]
> ```
> ### Near-Realtime Latency
> 
> Some internet applications that incorporate media streaming have specific interactivity or
> control-feedback requirements that drive much lower glass to glass media latency targets
> than one second.  These include videoconferencing or voice calls, remote video game play,
> remote control of hardware platforms like drones, vehicles, or surgical robots, and many
> other envisioned or deployed interactive applications. [refs if we have them]
> 
> Applications with latency targets in these regimes are out of scope for this document.
> ```

I like this proposal.
Even though those latencies are out of the scope of the doc, it is not explicitly clear.
Furthermore, the label "ultra-low" leaves a feeling lower latencies are unimaginable, but there are a lot of use cases where much lower latencies are targeted. Mentioning those just to say lower latencies do exist, they can be generally named "near-realtime" or "interactive", but they are out of the scope would be extremely helpful!

**Comment by acbegen** [comment]
I am personally getting tired of this topic coming up again and again. No hard feelings against any proposal but this is the exact reason why we wanted to avoid this discussion in the first place. I will have a look at the latest text proposal but in principle, I still don't like the text getting into latency categories not relevant for streaming media.

File: draft-ietf-mops-streaming-opcons.md (hunk 1 above)
Comment (acbegen) [reply draft-ietf-mops-streaming-opcons.md:562]: I am fine with adding some text @GrumpyOldTroll after our current categories and just before Section 4.1, but I am *not* OK to add another category to the list named "near realtime latency (less than 200 milliseconds)" or by any other name to that effect. This is a can of worms, as soon as we out 200ms in there, someone else will start complaining about why not 50 ms. We are obviously not denying the need for such latencies but adding that category is not going to help anybody. 

New text should be short and crisp, something along the lines of:
We note that some emerging applications and services _(e.g., list a few examples if desired)_ may need latencies in the order of a few hundreds of milliseconds or even less driven by interactivity requirements. However, the specific use cases and techniques for satisfying these requirements is out of scope for this document.

Comment (GrumpyOldTroll) [reply draft-ietf-mops-streaming-opcons.md:562]: I agree we should not add a category.  I also agree it's a can of worms and the main point of this discussion is to close it better so that people stop asking about it when they read this bit.

I do think it would be useful to add a sub-section under ultra-low latency to break it away from the rest of the text there, since it's a bit long already.  I made a [new PR](https://github.com/ietf-wg-mops/draft-ietf-mops-streaming-opcons/compare/issue-130) from my [latest proposed text](https://github.com/ietf-wg-mops/draft-ietf-mops-streaming-opcons/pull/129#issuecomment-1110513354) in the long thread.  I was just aiming to do that separation and clarify the "Applications with further-specialized latency requirements" in our out-of-scope statement a little more specifically.

I don't mind changing it to your text if you like it better, but I thought putting it at the end in place of our existing "out of scope latency" statement was a smaller change.

**Comment by maxsharabayko** [comment]
> This document specifically focuses on streaming applications and
   defines streaming as follows:
>   *  Streaming is transmission of a continuous media from a server to a
      client and its simultaneous consumption by the client.
>   *  Here, "continuous media" refers to media and associated streams
      such as video, audio, metadata, etc.  In this definition, the
      critical term is "simultaneous", as it is not considered streaming
      if one downloads a video file and plays it after the download is
      completed, which would be called download-and-play.

Sorry for bringing this up, @acbegen  🙂 
After this discussion my personal impression as an external (non-MOPS) reader of the document, that maybe the definition of the scope could be improved a bit then.
I honestly read the document several times and don't see why e.g. telesurgery does not fall into the category of "streaming media" as defined by this doc.
Although now after your comments I've got the feeling that this document defines "streaming" as maybe something with intermediate live processing or maybe more general-purpose streaming of some live events. **Presumably unidirectional streaming.** Not live streaming from e.g. military drones.
I feel like I am still missing some explicit definition of "streaming media" that would make it obvious that those near-realtime latencies are out of the scope.

Feel free to close the PR if you think defining lower latencies does not bring much.

**Comment by acbegen** [comment]
> > This document specifically focuses on streaming applications and
> > defines streaming as follows:
> > 
> > * Streaming is transmission of a continuous media from a server to a
> >   client and its simultaneous consumption by the client.
> > * Here, "continuous media" refers to media and associated streams
> >   such as video, audio, metadata, etc.  In this definition, the
> >   critical term is "simultaneous", as it is not considered streaming
> >   if one downloads a video file and plays it after the download is
> >   completed, which would be called download-and-play.
> 
> Sorry for bringing this up, @acbegen 🙂 After this discussion my personal impression as an external (non-MOPS) reader of the document, that maybe the definition of the scope could be improved a bit then. I honestly read the document several times and don't see why e.g. telesurgery does not fall into the category of "streaming media" as defined by this doc. 

Hi Maxim, telesurgery type of stuff requires interactivity. What you do with your robotic arm is important in that scenario, unlike what you do with your remote controller or mouse when you watch Netflix. We can give a more specific definition (that might make this clearer) but I felt like it would be too "heavy" right at the very beginning of the document. Drone videos, I would consider real-time video. And the fact that we don't list any other explicit categories does not mean they are excluded. It only means we don't make further detailed classification because again remote surgery vs. drone vs. twitch vs. online gaming are still quite a bit different.

**Comment by SpencerDawkins** [comment]
@maxsharabayko - we went a different way, with PR #136 - but this discussion was very helpful. Thanks for engaging!
